sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/OnTimerInvokersTest.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.Timer;
+import org.apache.beam.sdk.util.TimerSpec;
+import org.apache.beam.sdk.util.TimerSpecs;
+import org.apache.beam.sdk.util.WindowingInternals;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link DoFnInvokers}. */
+@RunWith(JUnit4.class)
+public class OnTimerInvokersTest {
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  @Mock private BoundedWindow mockWindow;
+
+  private ExtraContextFactory<String, String> extraContextFactory;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    this.extraContextFactory =
+        new ExtraContextFactory<String, String>() {
+          @Override
+          public BoundedWindow window() {
+            return mockWindow;
+          }
+
+          @Override
+          public DoFn.InputProvider<String> inputProvider() {
+            throw new UnsupportedOperationException("Not supposed to call this.");
+          }
+
+          @Override
+          public DoFn.OutputReceiver<String> outputReceiver() {
+            throw new UnsupportedOperationException("Not supposed to call this.");
+          }
+
+          @Override
+          public WindowingInternals<String, String> windowingInternals() {
+            throw new UnsupportedOperationException("Not supposed to call this.");
+          }
+
+          @Override
+          public <RestrictionT> RestrictionTracker<RestrictionT> restrictionTracker() {
+            throw new UnsupportedOperationException("Not supposed to call this.");
+          }
+        };
+  }
+
+  private void invokeOnTimer(DoFn<String, String> fn, String timerId) {
+    OnTimerInvokers.INSTANCE
+        .invokerForTimer(fn, timerId)
+        .invokeOnTimer(extraContextFactory);
+  }
+
+  @Test
+  public void testOnTimerHelloWord() throws Exception {
+    final String TIMER_ID = "my-timer-id";
+
+    class SimpleTimerDoFn extends DoFn<String, String> {
+
+      public String status = "not yet";
+
+      @TimerId(TIMER_ID)
+      private final TimerSpec myTimer = TimerSpecs.timer(TimeDomain.PROCESSING_TIME);
+
+      @ProcessElement
+      public void process(ProcessContext c) {}
+
+      @OnTimer(TIMER_ID)
+      public void onMyTimer() {
+        status = "OK now";
+      }
+    }
+
+    SimpleTimerDoFn fn = new SimpleTimerDoFn();
+
+    invokeOnTimer(fn, TIMER_ID);
+    assertThat(fn.status, equalTo("OK now"));
+  }
+
+  @Test
+  public void testOnTimerWithWindow() throws Exception {
+
+    WindowedTimerDoFn fn = new WindowedTimerDoFn();
+
+    invokeOnTimer(fn, WindowedTimerDoFn.TIMER_ID);
+    assertThat(fn.window, theInstance(mockWindow));
+  }
+
+  private static class WindowedTimerDoFn extends DoFn<String, String> {
```

Comment:
I have only tested (1) that it works to invoke the method and (2) that it can pass _some_ parameter. More is obviously called for. I hope to make the parameter passing code shared so obviously that it does not require independent coverage.